eclass/dlang-compilers-r1.eclass

```diff
@@ -0,0 +1,291 @@
+# two keywords.
+readonly _DLANG_LDC2_FRONTENDS=(
+	"1.32 2.102 ~amd64 ~arm64 ~x86"
+	"1.33 2.103 ~amd64 ~arm64 ~x86"
+	"1.34 2.104 ~amd64 ~arm64 ~x86"
+	"1.35 2.105 ~amd64 ~arm64 ~x86"
+	"1.36 2.106 ~amd64 ~arm64 ~x86"
+)
+
+# @FUNCTION: _dlang_accumulate_implementations
+# @INTERNAL
+# @DESCRIPTION:
+# Set the global variable _DLANG_ALL_IMPLS based on the three arrays:
+# _DLANG_(DMD|LDC2|GDC)_FRONTENDS. This function should be called once,
+# in global scope.
+_dlang_accumulate_implementations() {
+	local line result=()
+	for line in "${_DLANG_DMD_FRONTENDS[@]/#/dmd-}" \
+				"${_DLANG_GDC_FRONTENDS[@]/#/gdc-}" \
+				"${_DLANG_LDC2_FRONTENDS[@]/#/ldc2-}"
+	do
+		line=( ${line} )
+		# We only need the first component (name + version)
+		local impl=${line[0]/\./_}
+		result+=( "${impl}" )
+	done
+
+	if [[ ${_DLANG_ALL_IMPLS+1} ]]; then
+		if [[ ${_DLANG_ALL_IMPLS[@]} != ${result[@]} ]]; then
+			eerror "_DLANG_ALL_IMPLS has changes between inherits!"
+			eerror "Before: ${_DLANG_ALL_IMPLS[*]}"
+			eerror "Now   : ${result[*]}"
+			die "_DLANG_ALL_IMPLS integrity check failed!"
+		fi
+	else
+		_DLANG_ALL_IMPLS=( "${result[@]}" )
+		readonly _DLANG_ALL_IMPLS
+	fi
+}
+_dlang_accumulate_implementations
+unset -f _dlang_accumulate_implementations
+
+# @ECLASS_VARIABLE: _DLANG_ALL_IMPLS
+# @INTERNAL
+# @DESCRIPTION:
+# All supported Dlang implementations, most preferred last.
+
+# @ECLASS_VARIABLE: _DLANG_HISTORICAL_IMPLS
+# @INTERNAL
+# @DESCRIPTION:
+# All historical Dlang implementations that are no longer supported.
+readonly _DLANG_HISTORICAL_IMPLS=(
+	gdc-11
+)
+
+# @FUNCTION: dlang_compilers_stabs_required_use
+# @USAGE: <use_expand_name> [<impls>...]
+# @DESCRIPTION:
+# Generate a required-use expression which ensures that only supported
+# implementations can be selected.
+#
+# Check _dlang_get_bad_arches for a more detailed description of what
+# an implementation being supported means.
+#
+# <use_expand_name> is a string used to transform implementations as
+# they appear in <impls> into USE flags. Possible values are
+# "dlang_targets" and "dlang_single_target".
+#
+# Example: If we had the implementation dmd-2_105 with the keywords,
+# i.e. the keywords in _DLANG_DMD_FRONTENDS, `amd64 ~x86' and $KEYWORDS
+# has the contents `amd64 arm64 x86' the output should be:
+# @CODE
+# "<use_expand_name>_dmd-2_105? ( !x86 ) <use_expand_name>_dmd-2_105? ( !arm64 )"
+# @CODE
+dlang_compilers_stabs_required_use() {
+	debug-print-function ${FUNCNAME} "${@}"
+
+	local use_expand_name=${1}
+	local impls=( ${@:2} )
+
+	local result=() impl
+	for impl in "${impls[@]}"; do
+		local badArches=( $(_dlang_get_bad_arches "${impl}") )
+		local arch
+		for arch in "${badArches[@]}"; do
+			result+=( "${use_expand_name}_${impl}? ( !${arch} )" )
+		done
+	done
+
+	echo "${result[@]}"
+}
+
+# @FUNCTION: dlang_compilers_stabs_impl_dep
+# @USAGE: <impl> <dep_string>
+# @DESCRIPTION:
+# Print a dependency string based on <dep_string>, USE disabled on
```

Comment:
This again feels kind of scary and I don't think we should be doing this at all...

Reply:
Ok, I've had plenty of time to consider the implications of doing this, the ups and downs and possible solutions so let me share it:

1. My aim with the new eclasses was to accomplish the same goal as the old eclass, just in a more standard and readable way. This means that I've kept what the eclass was doing in regards to the KEYWORDS abuse.
2. I understand that using the KEYWORDS variable is bad stuff and shouldn't be done and it isn't done in ::gentoo, because all of this functionality is meant to be expressed through profiles, which I can't do here without having users explicitly use profiles in this overlay.
3. I don't know how removing these 2 functions would affect the general user experience. The most common interaction that these additions affect is trying to install a stable package with an unstable compiler, in which the KEYWORDS abuse implementation gives a cryptic REQUIRED_USE failure, and the do-nothing implementation just allows it so, I guess, it's a win for doing nothing?
4. the developer experience however, is much worse. Here's what pkgcheck output if those USE flags are not generated on a properly stabilized dub:
```
dev-util/dub
  RequiredUseDefaults: version 1.36.0-r1: profile: 'default/linux/amd64/17.0/musl' (132 total) failed REQUIRED_USE: exactly-one-of ( dlang_single_target_dmd-2_101 dlang_single_target_dmd-2_102 dlang_single_target_dmd-2_103 dlang_single_target_dmd-2_104 dlang_single_target_dmd-2_105 dlang_single_target_dmd-2_106 dlang_single_target_dmd-2_107 dlang_single_target_gdc-12 dlang_single_target_gdc-13 dlang_single_target_ldc2-1_32 dlang_single_target_ldc2-1_33 dlang_single_target_ldc2-1_34 dlang_single_target_ldc2-1_35 dlang_single_target_ldc2-1_36 )
  MissingUnpackerDep: version 1.36.0-r1: missing BDEPEND="app-arch/unzip" for SRC_URI archives: [ gitcompatibledubpackage-1.0.1.zip, gitcompatibledubpackage-1.0.4.zip, urld-2.1.1.zip ]
  NonsolvableDepsInDev: version 1.36.0-r1: nonsolvable depset(bdepend) keyword(amd64) dev profile (default/linux/amd64/17.0/musl) (8 total): solutions: [ dev-lang/dmd:2.107=, dev-lang/ldc2:1.36= ]
  NonsolvableDepsInDev: version 1.36.0-r1: nonsolvable depset(bdepend) keyword(~arm64) dev profile (default/linux/arm64/17.0/hardened/selinux) (4 total): solutions: [ dev-lang/dmd:2.101=, dev-lang/dmd:2.102=, dev-lang/dmd:2.103=, dev-lang/dmd:2.104=, dev-lang/dmd:2.105=, dev-lang/dmd:2.106=, dev-lang/dmd:2.107= ]
  NonsolvableDepsInDev: version 1.36.0-r1: nonsolvable depset(depend) keyword(amd64) dev profile (default/linux/amd64/17.0/musl) (8 total): solutions: [ dev-lang/dmd:2.107=, dev-lang/ldc2:1.36= ]
  NonsolvableDepsInDev: version 1.36.0-r1: nonsolvable depset(depend) keyword(~arm64) dev profile (default/linux/arm64/17.0/hardened/selinux) (4 total): solutions: [ dev-lang/dmd:2.101=, dev-lang/dmd:2.102=, dev-lang/dmd:2.103=, dev-lang/dmd:2.104=, dev-lang/dmd:2.105=, dev-lang/dmd:2.106=, dev-lang/dmd:2.107= ]
  NonsolvableDepsInDev: version 1.36.0-r1: nonsolvable depset(rdepend) keyword(amd64) dev profile (default/linux/amd64/17.0/musl) (8 total): solutions: [ dev-lang/dmd:2.107=, dev-lang/ldc2:1.36= ]
  NonsolvableDepsInDev: version 1.36.0-r1: nonsolvable depset(rdepend) keyword(~arm64) dev profile (default/linux/arm64/17.0/hardened/selinux) (4 total): solutions: [ dev-lang/dmd:2.101=, dev-lang/dmd:2.102=, dev-lang/dmd:2.103=, dev-lang/dmd:2.104=, dev-lang/dmd:2.105=, dev-lang/dmd:2.106=, dev-lang/dmd:2.107= ]
  NonsolvableDepsInStable: version 1.36.0-r1: nonsolvable depset(bdepend) keyword(amd64) stable profile (default/linux/amd64/17.1) (48 total): solutions: [ dev-lang/dmd:2.107=, dev-lang/ldc2:1.36= ]
  NonsolvableDepsInStable: version 1.36.0-r1: nonsolvable depset(bdepend) keyword(~arm64) stable profile (default/linux/arm64/17.0) (16 total): solutions: [ dev-lang/dmd:2.101=, dev-lang/dmd:2.102=, dev-lang/dmd:2.103=, dev-lang/dmd:2.104=, dev-lang/dmd:2.105=, dev-lang/dmd:2.106=, dev-lang/dmd:2.107= ]
  NonsolvableDepsInStable: version 1.36.0-r1: nonsolvable depset(depend) keyword(amd64) stable profile (default/linux/amd64/17.1) (48 total): solutions: [ dev-lang/dmd:2.107=, dev-lang/ldc2:1.36= ]
  NonsolvableDepsInStable: version 1.36.0-r1: nonsolvable depset(depend) keyword(~arm64) stable profile (default/linux/arm64/17.0) (16 total): solutions: [ dev-lang/dmd:2.101=, dev-lang/dmd:2.102=, dev-lang/dmd:2.103=, dev-lang/dmd:2.104=, dev-lang/dmd:2.105=, dev-lang/dmd:2.106=, dev-lang/dmd:2.107= ]
  NonsolvableDepsInStable: version 1.36.0-r1: nonsolvable depset(rdepend) keyword(amd64) stable profile (default/linux/amd64/17.1) (48 total): solutions: [ dev-lang/dmd:2.107=, dev-lang/ldc2:1.36= ]
  NonsolvableDepsInStable: version 1.36.0-r1: nonsolvable depset(rdepend) keyword(~arm64) stable profile (default/linux/arm64/17.0) (16 total): solutions: [ dev-lang/dmd:2.101=, dev-lang/dmd:2.102=, dev-lang/dmd:2.103=, dev-lang/dmd:2.104=, dev-lang/dmd:2.105=, dev-lang/dmd:2.106=, dev-lang/dmd:2.107= ]
```
I'm my opinion that is way too bad a trade-off for me as a developer to have to go through that list and see if there's an actual issue that I should fix or if it's just the normal amount or errors.
5. Maybe some of this can be alleviated if ::dlang stopped adding stabilization keywords, or at least it would get around the most common (I believe) issue users face with the overlay (the REQUIRED_USE failure), but this decision has consequences and I don't think I have the experience to make it.

Given all of this, did your stance change on the issue? Is there anything that I can do to make this better?